src/input/gamepad/gamepadCalibration.js

```diff
@@ -0,0 +1,500 @@
+// @flow
+/*eslint indent:0*/
+
+import {Vec2D} from "../../main/util/Vec2D";
+import {deepCopyObject} from "../../main/util/deepCopyObject";
+import {setCustomGamepadInfo} from "./gamepads/custom";
+import {setUsingCustomControls, currentPlayers} from "../../main/main";
+import {updateControllerMenu} from "../../menus/controllermenu.js";
+import {nullGamepadInfo} from "./gamepadInfo";
+import {getGamepad} from "./gamepad";
+
+// eslint-disable-next-line no-duplicate-imports
+import type {ButtonInfo, StickInfo, StickCardinals, TriggerInfo, DPadInfo, GamepadInfo} from "./gamepadInfo";
+// eslint-disable-next-line no-duplicate-imports
+import type {Gamepad, Button} from "./gamepad";
+
+
+const calibrationInProgress = [false, false, false, false];
+function setCalibrationInProgress (i : number, bool : bool) : void {
+  calibrationInProgress[i] = bool;
+}
+
+function deepCopyButtons( gpd : Gamepad ) : Array<Button> {
+  const lgB = gpd.buttons.length;
+  const buttons = [];
+  for (let i = 0; i < lgB; i++) {
+    const button = gpd.buttons[i];
+    buttons.push({value : button.value, pressed : button.pressed});
+  } 
+  return buttons;
+}
+
+function deepCopyAxes( gpd : Gamepad ) : Array<number> {
+  const lgA = gpd.axes.length;
+  const axes = [];
+  for (let j = 0; j < lgA; j++) {
+    axes.push(gpd.axes[j]);
+  }
+  return axes;
+}
+
+type Snapshots = { b0 : Array<Button>, bL : Array<Button>, bR : Array<Button>, bU : Array<Button>
+                 , a0 : Array<number>, aL : Array<number>, aR : Array<number>, aU : Array<number> };
+
+const nullSnapshots : Snapshots = { b0 : [], bL : [], bR : [], bU : []
+                                  , a0 : [], aL : [], aR : [], aU : [] };
+
+
+export function runCalibration ( i : number ) : void {
+  if (calibrationInProgress[i]) {
+    return;
+  }
+  setCalibrationInProgress(i, true);
+  
+  const interval = 3000; // 3 seconds
+
+  const j = currentPlayers[i];
+
+  const gamepadInfo : GamepadInfo = nullGamepadInfo;
+  gamepadInfo.ids = [ { name : "custom controller" } ];
+  calibrateGamepad( i, j, gamepadInfo
+                  , nullSnapshots
+                  , 0, interval
+                  );
+
+}
+
+function calibrateGamepad ( i : number, j : number
+                          , gamepadInfo : GamepadInfo
+                          , snapshots : Snapshots
+                          , passNumber : number, interval : number ) : void {
+  let text;
+  let gamepad;
+
+  switch(passNumber) { // fire events
```

Comment:
theres a lot of duplication here. I believe i've got a trick or two for this. ill get back to you.